### ISSUE TITLE
The method 'move' was called on null.

### ISSUE THREAD

**Issue by Skyost** [johnpryan/flutter_map#337]
Here is the stacktrace :

```dart
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following NoSuchMethodError was thrown building StreamBuilder<LatLng>(dirty, state:
_StreamBuilderBaseState<LatLng, AsyncSnapshot<LatLng>>#d59a1):
The method 'move' was called on null.
Receiver: null
Tried calling: move(Instance of 'LatLng', 18.0, hasGesture: false, isUserGesture: false)
When the exception was thrown, this was the stack:
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:50:5)
#1      MapControllerImpl.move (package:flutter_map/src/map/map.dart:28:12)
#2      SelfUpdatingMap.build.<anonymous closure> (package:car_park/util/self_updating_map.dart:58:22)
#3      StreamBuilder.build (package:flutter/src/widgets/async.dart:425:74)
#4      _StreamBuilderBaseState.build (package:flutter/src/widgets/async.dart:125:48)
#5      StatefulElement.build (package:flutter/src/widgets/framework.dart:3825:27)
#6      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3739:15)
#7      Element.rebuild (package:flutter/src/widgets/framework.dart:3565:5)
#8      BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2278:33)
#9      _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:700:20)
#10     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:286:5)
#11     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1012:15)
#12     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:952:9)
#13     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:864:5)
#17     _invoke (dart:ui/hooks.dart:219:10)
#18     _drawFrame (dart:ui/hooks.dart:178:3)
(elided 3 frames from package dart:async)
════════════════════════════════════════════════════════════════════════════════════════════════════
```

In fact, I get this error when I call `move(...)` on a `MapController` which is not affected to a map.

**Comment by kaynz** [comment]
Same here with latest flutter and flutter_map 0.5.4

**Comment by Skyost** [comment]
Okay, the usage of `MapController` is not well documented but the class has a getter `ready` which returns whether or not the state can be used. So before using any `MapController` method, the `ready` boolean should be checked @kaynz.

**Comment by MitchellQ** [comment]
@Skyost have you got this working successfully because I am having the same issue except that center and zoom also are null so I can't even reach map.move 

any help getting this working would be appreciated

**Comment by Skyost** [comment]
@MitchellQ 

You have to wait until `controller.ready` returns _true_ before using the methods. You may want to use `controller.onReady` to know when you can use the methods like `move`, `center`, etc...

**Comment by MitchellQ** [comment]
I'm not having much luck @Skyost can you please help
```
 mapController.onReady.then((result) {
      setState(() {
        animatedMapMove(
          LatLng(tracks[trackNum].activities[0].location.x,
              tracks[trackNum].activities[0].location.y),
          18.0,
        );
      });
    });
```

I changed my code to use onReady but it just skips over and never goes back, Im quite new to flutter and developing in general so I'm probably missing something simple but any help is appreciated

**Comment by Skyost** [comment]
@MitchellQ Try something like that :

```dart
mapController.onReady.then((result) {
  mapController.move(LatLng(tracks[trackNum].activities[0].location.x, tracks[trackNum].activities[0].location.y), 18.0);
});
```

**Comment by MitchellQ** [comment]
@Skyost I got it working, I was able to get it working and didn't need to use onReady. It seems my problem was due to using bottom navigation and how i displayed the map namely I have a List of widgets that the map is part of and the animate method doesn't like that for some reason so I used a ternary to display the map on the tab instead of adding it to the List of Widgets I spent so many hours on this, thanks for your help

**Comment by ascmp** [comment]
I face a similar issue with this snapshot of code:
```
  @override
  Widget build(BuildContext context) {
    var center = getCenter(widget.issues);
    if (_mapCtrl.ready) {
      _mapCtrl.move(center, 12.0);
    }
```
and the stacktrace is:
```
I/flutter (23365): ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
I/flutter (23365): The following NoSuchMethodError was thrown building LeafletMap(dirty, state:
I/flutter (23365): _LeafletMapState#2fe39):
I/flutter (23365): The method 'toDouble' was called on null.
I/flutter (23365): Receiver: null
I/flutter (23365): Tried calling: toDouble()
I/flutter (23365): 
I/flutter (23365): When the exception was thrown, this was the stack:
I/flutter (23365): #0      Object.noSuchMethod (dart:core-patch/object_patch.dart:50:5)
I/flutter (23365): #1      pow (dart:math-patch/math_patch.dart:83:44)
I/flutter (23365): #2      Crs.scale (package:flutter_map/src/geo/crs/crs.dart:38:18)
I/flutter (23365): #3      MapState.getZoomScale (package:flutter_map/src/map/map.dart:237:16)
I/flutter (23365): #4      MapState.getPixelBounds (package:flutter_map/src/map/map.dart:261:17)
I/flutter (23365): #5      MapState.getLastPixelBounds (package:flutter_map/src/map/map.dart:165:12)
I/flutter (23365): #6      MapState._calculateBounds (package:flutter_map/src/map/map.dart:169:18)
I/flutter (23365): #7      MapState.getBounds (package:flutter_map/src/map/map.dart:157:12)
I/flutter (23365): #8      MapState.bounds (package:flutter_map/src/map/map.dart:86:30)
I/flutter (23365): #9      MapState.move (package:flutter_map/src/map/map.dart:103:56)
I/flutter (23365): #10     MapControllerImpl.move (package:flutter_map/src/map/map.dart:28:12)
```
any help is appreciate on what I could possibly do wrong

**Comment by johnpryan** [comment]
@Skyost is correct, the mapController can't be used until it's ready

**Comment by Maadhav** [comment]
It is still not working. Please help me I tried everything from above.

**Comment by epsi95** [comment]
As other said, we need to check onReady
`floatingActionButton: FloatingActionButton(
        onPressed: () {
          _mapController.onReady.then((result) {
            _mapController.move(LatLng(51.5, -0.09), 13.0);
          });
        },
      )`

At first, I was facing issues because _mapController was initialized at build() method, so as soon as I was doing setState(), old _mapController.onReady was getting discarded. Make sure _mapController is declared at class level. Now everything works fine.

**Comment by AldoG94** [comment]
if (lat != null && long != null) {
      mapController.onReady.then((result) {
        mapController.move(LatLng(lat, long), _inZoom);
This is part of my code.
I hope it works for u guys! it works for me
I put in a function where move the cam if the location (lat,long) is already to work

**Comment by cedvdb** [comment]
My map controller is never ready. Any idea why ?

**Comment by ibrierley** [comment]
Not without seeing any code no. It may be that you're drawing the widget before the map controller is ready as mentioned earlier.

**Comment by hieudz** [comment]
> My map controller is never ready. Any idea why ?

My map controller is never ready too. When I create a global map controller, it works normally, but when I create locally it doesn't work

**Comment by ibrierley** [comment]
Can you include some code with the mapcontroller defined and where it's set.

**Comment by hieudz** [comment]
I'm not sure how to include the code in a concise way. I'm using flutter_bloc to control my states. I work normally before I insert the bloc builder, but after I include that I cannot make it work. Also If I replace the `_mapcontroller` initiation with a global _mapcontroller, it works normally.
```
return BlocBuilder<AppMapCubit, AppMapState>(
      builder: (context, appMapState) {
        MapController _mapController = MapController();
        AppMapStateEnabled state = appMapState;
        if(state.userLocation == null)
          return Loading();

        return Stack(children: [
          FlutterMap(
            children: [
              Builder(
                  builder: (context) {
                    MapState.of(context).rebuildLayers();
                    return Container();
                  }
              )
            ],
            mapController: _mapController,
            options: MapOptions(
              center: LatLng(
                  state.userLocation.latitude,
                  state.userLocation.longitude
              ),
              zoom: 16,
            ),
            layers: [
              TileLayerOptions(
                urlTemplate: "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
                subdomains: ['a', 'b', 'c'],
                tileProvider: NetworkTileProvider(),
              ),
              CircleLayerOptions(
                circles: [
                  _circleMarkerForUser(state.partnerLocation, 1),
                ],
              ),
              MarkerLayerOptions(
                markers: [
                  _markerForUser(state.userLocation, 0),
                  _markerForUser(state.partnerLocation, 1),
                ],
              ),
            ],
          ),
          Positioned(
            child: FloatingActionButton(
              heroTag: "btn3",
              onPressed: () {
                print("So we have $_mapController");
                _mapController.onReady.then(
                  (_) => _mapController.move(
                    LatLng(state.userLocation.latitude,
                        state.userLocation.longitude),
                    16
                  )
                );
              },
              child: const Icon(Icons.my_location),
            ),
            bottom: 20,
            right: 20,
          ),
       ]);
      }
    );
  }
```

**Comment by ibrierley** [comment]
I suspect you want your mapcontroller defined and assigned outside of the builder, otherwise I think you will be creating a new one every state change which doesn't make sense. Apologies I'm not an expert on Bloc stuff, but most builders are called on a certain change, but typically you don't want to create a new instance every state change here for the mapcontroller.

**Comment by hieudz** [comment]
Actually I tried that before and it show the exact same error so I try to put it in the builder. Do you have any idea on what can block a map controller from being ready?

**Comment by ibrierley** [comment]
Can you post the code with the mapController assigned outside of the Blockbuilder and the error it gives.

**Comment by hieudz** [comment]
The code is exactly the same as above, except for the MapController is created right before I return the widget. It also doesn't produce any error because the MapController has never been ready. If I remove the onReady and try `_mapController.move(...) `instead, it raise the error
```

The following NoSuchMethodError was thrown while handling a gesture:
The method 'move' was called on null.
Receiver: null
Tried calling: move(Instance of 'LatLng', 16.0, id: null, source: Instance of 'MapEventSource')

When the exception was thrown, this was the stack: 
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:51:5)
#1      MapControllerImpl.move (package:flutter_map/src/map/map.dart:41:19)
#2      _MapViewState.build.<anonymous closure>.<anonymous closure> (package:longggglove/screens/map/map_view.dart:153:32)
#3      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:993:19)
#4      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:1111:38)
...
Handler: "onTap"
Recognizer: TapGestureRecognizer#17329
  debugOwner: GestureDetector
  state: possible
  won arena
  finalPosition: Offset(365.0, 809.6)
  finalLocalPosition: Offset(29.6, 31.9)
  button: 1
  sent tap down
==================================================================================================== 
```

**Comment by hieudz** [comment]
The weird thing is that the map controller will work whenever I try to put mapcontroller outside of the build function (the function that return this one https://github.com/fleaflet/flutter_map/issues/337#issuecomment-803054166)

**Comment by ibrierley** [comment]
Why is it weird ? That's how its supposed to be, you don't want to be setting up a controller every time build is called. Set up the controller when the state is initialised.

**Comment by Ryadhmd** [comment]
My mapcontroller doesn't get ready too, and I don't know why, it not in the build method though

**Comment by ibrierley** [comment]
Feel free to post some code.

**Comment by Ryadhmd** [comment]
okay here is the build method of my home state  : 
`@override
  Widget build(BuildContext context) {
  
    return FutureBuilder(
        future: chargerMonument(),
        builder: (context, snaphot) {
          if (snaphot.hasData) {
            return SafeArea(
      
              child: Scaffold(
                appBar: AppBar(
                        actions: [
                          IconButton(
                              icon: Icon(Icons.search),
                              onPressed: () {
                                showSearch(
                                    context: context,
                                    delegate: CustomSearchDelegate(
                                        monuments, updatelocation));
                              }),
                        ],
                   
                key: _scaffoldKey,
  
                body: Map(),`

for Map : 

`class Map extends StatefulWidget {
  @override
  _MapPageState createState() => _MapPageState();
}

class _MapPageState extends State<Map> {
  static const _markerSize = 30.0;
  List<Marker> _markers;
  List<LatLng> _points;
  List<LatLng> polylines;
  double distance;
  final MapController mapController=MapController();
 CenterOnLocationUpdate _centerOnLocationUpdate;
  StreamController<double> _centerCurrentLocationStreamController;
final PopupController _popupLayerController = PopupController();


 
  @override
  void initState() {
    super.initState();

   
    _centerOnLocationUpdate = CenterOnLocationUpdate.always;
    _centerCurrentLocationStreamController = StreamController<double>();
  }

  
  @override
  void dispose() {
    _centerCurrentLocationStreamController.close();
    super.dispose();
  }

  Future<String> test() async {
    QuerySnapshot<Object> querySnapshot = await DatabaseService().getData();
    _points = DatabaseService().convert(querySnapshot);
    _markers = _points
        .map(
          (LatLng point) => Marker(
              point: point,
              builder: (_) => Icon(Icons.location_on, size: _markerSize),
              width: _markerSize,
              height: _markerSize,
              anchorPos: AnchorPos.align(AnchorAlign.top)),
        )
        .toList();
    return 'Okay';
  }

  void drawpolyline(List<LatLng> line, String desti, double d, int refreash) {
    print(line.length.toString());
    setState(() {
      polylines = line;
      destination = desti;
      distance = d;
      refresh = refresh;
      _centerOnLocationUpdate = CenterOnLocationUpdate.always;
      _centerCurrentLocationStreamController.add(18);
    });
  }

  @override
  Widget build(BuildContext context) {

    return /* TimerBuilder.periodic(Duration(seconds: refresh),
        builder: (context) => */ FutureBuilder(
            future: test(),
            builder: (BuildContext context, AsyncSnapshot<String> snapshot) {
              List<Widget> children;
              if (snapshot.hasData) {
                children = [
                  Flexible(
                    child: FlutterMap(
                      children: [
                        LocationMarkerLayerWidget(
                          plugin: LocationMarkerPlugin(
                              centerCurrentLocationStream:
                                  _centerCurrentLocationStreamController.stream,
                              //_centerOnLocationUpdate
                              centerOnLocationUpdate: _centerOnLocationUpdate),
                        )
                      ],
                      options: MapOptions(
                        onPositionChanged:
                            (MapPosition position, bool hasGesture) {
                          if (hasGesture) {
                            setState(() => _centerOnLocationUpdate =
                                CenterOnLocationUpdate.never);
                          }
                        },
                        plugins: [
                          PopupMarkerPlugin(),
                          ZoomButtonsPlugin(),
                          LocationPlugin(),
                          TappablePolylineMapPlugin()
                        ],
                        interactiveFlags:
                            InteractiveFlag.all & ~InteractiveFlag.rotate,
                        zoom: 18.0,
                        center: LatLng(0, 0),
                        onTap: (_) => _popupLayerController
                            .hidePopup(), // Hide popup when the map is tapped.
                      ),
                      layers: [
                        TileLayerOptions(
                          urlTemplate:
                              'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
                          subdomains: ['a', 'b', 'c'],
                        ),
                        TappablePolylineLayerOptions(
                            // Will only render visible polylines, increasing performance
                            polylineCulling: true,
                            pointerDistanceTolerance: 20,
                            polylines: [
                              TaggedPolyline(
                                  points: polylines,
                                  isDotted: true,
                                  color: Colors.blue[300],
                                  strokeWidth: 4.0)
                            ],
                            onTap: (TaggedPolyline polyline) async {
                              await showDialog(
                                  context: context,
                                  builder: (_) => new AlertDialog(
                                          title: Text(destination),
                                          content: Text("Vous etes à " +
                                              distance.round().toString() +
                                              "km de votre distination !"),
                                          actions: [
                                            TextButton(
                                                onPressed: () {
                                                  Navigator.of(context).pop();
                                                },
                                                child: Text("D'accord !")),
                                            TextButton(
                                                onPressed: () {
                                                  setState(() {
                                                    polylines = [];
                                                    refresh = 360000;
                                                  });
                                                  Navigator.of(context).pop();
                                                },
                                                child:
                                                    Text("Effacer ce chemin"))
                                          ]));
                            },
                            onMiss: () {
                              print('No polyline was tapped');
                            }),
                        PopupMarkerLayerOptions(
                            markers: _markers,
                            popupSnap: PopupSnap.markerTop,
                            popupController: _popupLayerController,
                            popupBuilder: (BuildContext _, Marker marker) =>
                                ExamplePopup(
                                    marker, marker.point, drawpolyline)),
                        ZoomButtonsPluginOption(
                            minZoom: 4,
                            maxZoom: 22,
                            mini: true,
                            padding: 10,
                            alignment: Alignment.bottomRight),
                        LocationOptions(onLocationUpdate: (LatLngData ld) {
                          print(
                              'Location updated: ${ld.location} (accuracy: ${ld.accuracy})');
                              print("etat:"+mapController.ready.toString());
                        }, onLocationRequested: (LatLngData ld) {
                          if (ld == null) {
                            return;
                          }
                          mapController.move(ld.location, 16.0);
                        },
                            /* onLocationUpdate: (LatLngData ld) {},
                            onLocationRequested: (LatLngData ld) {
                              if (ld == null || ld.location == null) {
                                return;
                              }
                              /* return;
                              }
                              mapController?.move(ld.location, 16.0);*/
                              mapController.onReady.then((result) {
                                print("I AM READY");
                                mapController.move(ld.location, 16);
                              });
                              */

                            buttonBuilder: (BuildContext context,
                                ValueNotifier<LocationServiceStatus> status,
                                Function onPressed) {
                          return Align(
                              alignment: Alignment.bottomRight,
                              child: Padding(
                                  padding: const EdgeInsets.only(
                                      bottom: 14.0, right: 60.0),
                                  child: Container(
                                    height: 38,
                                    width: 38,
                                    child: FloatingActionButton(
                                        backgroundColor: Colors.grey,
                                        child: ValueListenableBuilder<
                                                LocationServiceStatus>(
                                            valueListenable: status,
                                            builder: (BuildContext context,
                                                LocationServiceStatus value,
                                                Widget child) {
                                              switch (value) {
                                                case LocationServiceStatus
                                                    .disabled:
                                                case LocationServiceStatus
                                                    .permissionDenied:
                                                case LocationServiceStatus
                                                    .unsubscribed:
                                                  return const Icon(
                                                    Icons.location_disabled,
                                                    color: Colors.white,
                                                  );
                                                  break;
                                                default:
                                                  return const Icon(
                                                    Icons.my_location,
                                                    color: Colors.white,
                                                  );
                                                  break;
                                              }
                                            }),
                                        onPressed: () {
                                          setState(() =>
                                              _centerOnLocationUpdate =
                                                  CenterOnLocationUpdate
                                                      .always);
                                          _centerCurrentLocationStreamController
                                              .add(18);
                                        }),
                                  )));
                        })
                      ],
                    ),
                  ),
                ];
              } else if (snapshot.hasError) {
                children = <Widget>[
                  Icon(
                    Icons.error_outline,
                    color: Colors.red,
                    size: 60,
                  ),
                  Padding(
                    padding: const EdgeInsets.only(top: 16),
                    child: Text('Error: ${snapshot.error}'),
                  )
                ];
              } else {
                children = <Widget>[
                  SizedBox(
                    child: CircularProgressIndicator(),
                    width: 60,
                    height: 60,
                  )
                ];
              }
              return Center(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: children,
                ),
              );
            });
`
Thank you for helping me

**Comment by ibrierley** [comment]
It's hard to follow with the formatting being a little messed in the paste, but it does look like you have a mapController.move(ld.location, 16.0);  without any null check. Maybe try adding a
```if( mapController != null )``` before it ?

**Comment by ibrierley** [comment]
I'd also be tempted to move your mapController initialisation into the initState() method.

**Comment by Ryadhmd** [comment]
yeah, that line shows the error but my problem is that I need to call move later in my  but the MapController never gets ready, I tried to put it in the initState but same thing :/

**Comment by ibrierley** [comment]
Are you actually passing in mapController as an option to FlutterMap widget ? Maybe I'm blind, but I can't see it...try adding mapController: mapController as a parameter.

**Comment by Ryadhmd** [comment]
you weren't at all wrong, I'm soory thank you

**Comment by ibrierley** [comment]
No problem, great you got it sorted and hopefully it may help someone else who has the same problem.

**Comment by Ryadhmd** [comment]
I have a one more question, do you have any idea about how can I from my appbar when a button is pressed to call change the _centeronlocationupdate to CenterOnLocationUpdate.never and do a move to a certain location ?

**Comment by LondonJammo** [comment]
> MitchellQ

I think this is the real reason people are having this issue, as am I still. It's to do with bottomNavigation and the state being swapped out and not being found when you try to access the controller. 

> I got it working, I was able to get it working and didn't need to use onReady. It seems my problem was due to using bottom navigation and how i displayed the map namely I have a List of widgets that the map is part of and the animate method doesn't like that for some reason so I used a ternary to display the map on the tab instead of adding it to the List of Widgets I spent so many hours on this, thanks for your help

I'm still trying to find a tidy fix which uses the bottomNav in its current setup without having to use a ternary switcher

**Comment by JaffaKetchup** [comment]
@LondonJammo Interestingly, I just found myself in this situation yesterday. In order to keep state when navigating, I ended up having to use `PageView`. I can post some code if you're interested?
